File: tests/dataset/job/compiler/fortran/q5.f90
program q5
  implicit none
  type :: CompanyType
    integer :: ct_id
    character(len=32) :: kind
  end type CompanyType
  type :: InfoType
    integer :: it_id
    character(len=32) :: info
  end type InfoType
  type :: TitleRec
    integer :: t_id
    character(len=32) :: title
    integer :: year
  end type TitleRec
  type :: MovieCompany
    integer :: movie_id
    integer :: company_type_id
    character(len=64) :: note
  end type MovieCompany
  type :: MovieInfo
    integer :: movie_id
    character(len=32) :: info
    integer :: info_type_id
  end type MovieInfo
  type(CompanyType) :: company_types(2)
  type(InfoType) :: info_types(1)
  type(TitleRec) :: titles(3)
  type(MovieCompany) :: movie_companies(3)
  type(MovieInfo) :: movie_infos(3)
  character(len=32) :: allowed(8)
  character(len=32) :: best_title
  logical :: first
  integer :: i,j,k,l,m,n

  company_types(1) = CompanyType(1,'production companies')
  company_types(2) = CompanyType(2,'other')

  info_types(1) = InfoType(10,'languages')

  titles(1) = TitleRec(100,'B Movie',2010)
  titles(2) = TitleRec(200,'A Film',2012)
  titles(3) = TitleRec(300,'Old Movie',2000)

  movie_companies(1) = MovieCompany(100,1,'ACME (France) (theatrical)')
  movie_companies(2) = MovieCompany(200,1,'ACME (France) (theatrical)')
  movie_companies(3) = MovieCompany(300,1,'ACME (France) (theatrical)')

  movie_infos(1) = MovieInfo(100,'German',10)
  movie_infos(2) = MovieInfo(200,'Swedish',10)
  movie_infos(3) = MovieInfo(300,'German',10)

  allowed = (/ 'Sweden','Norway','Germany','Denmark', &
               'Swedish','Denish','Norwegian','German' /)

  first = .true.
  do i=1,2
    if (trim(company_types(i)%kind) == 'production companies') then
      do j=1,3
        if (movie_companies(j)%company_type_id == company_types(i)%ct_id .and. &
            index(movie_companies(j)%note,'(theatrical)') > 0 .and. &
            index(movie_companies(j)%note,'(France)') > 0) then
          do k=1,3
            if (movie_infos(k)%movie_id == movie_companies(j)%movie_id) then
              do l=1,1
                if (movie_infos(k)%info_type_id == info_types(l)%it_id) then
                  do m=1,3
                    if (titles(m)%t_id == movie_companies(j)%movie_id .and. titles(m)%year > 2005) then
                      do n=1,8
                        if (trim(movie_infos(k)%info) == allowed(n)) then
                          if (first) then
                            best_title = titles(m)%title
                            first = .false.
                          else
                            if (titles(m)%title < best_title) best_title = titles(m)%title
                          end if
                          exit
                        end if
                      end do
                    end if
                  end do
                end if
              end do
            end if
          end do
        end if
      end do
    end if
  end do

  print '(A)', '[{"typical_european_movie":"'//trim(best_title)//'"}]'
end program q5

File: tests/dataset/job/compiler/py/q5.py
from __future__ import annotations
import dataclasses
import json


@dataclasses.dataclass
class Auto1:
    typical_european_movie: str

    def __getitem__(self, key):
        return getattr(self, key)


@dataclasses.dataclass
class Auto2:
    ct_id: int
    kind: str

    def __getitem__(self, key):
        return getattr(self, key)


@dataclasses.dataclass
class Auto3:
    it_id: int
    info: str

    def __getitem__(self, key):
        return getattr(self, key)


@dataclasses.dataclass
class Auto4:
    t_id: int
    title: str
    production_year: int

    def __getitem__(self, key):
        return getattr(self, key)


@dataclasses.dataclass
class Auto5:
    movie_id: int
    company_type_id: int
    note: str

    def __getitem__(self, key):
        return getattr(self, key)


@dataclasses.dataclass
class Auto6:
    movie_id: int
    info: str
    info_type_id: int

    def __getitem__(self, key):
        return getattr(self, key)


from typing import Any, TypeVar, Generic, Callable

T = TypeVar("T")
K = TypeVar("K")


def _min(v):
    if hasattr(v, "Items"):
        v = v.Items
    if not isinstance(v, list):
        raise Exception("min() expects list or group")
    vals = [it for it in v if it is not None]
    if not vals:
        return 0
    return min(vals)


def _query(src, joins, opts):
    items = [[v] for v in src]
    for j in joins:
        joined = []
        if j.get("right") and j.get("left"):
            matched = [False] * len(j["items"])
            for left in items:
                m = False
                for ri, right in enumerate(j["items"]):
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    matched[ri] = True
                    joined.append(left + [right])
                if not m:
                    joined.append(left + [None])
            for ri, right in enumerate(j["items"]):
                if not matched[ri]:
                    undef = [None] * (len(items[0]) if items else 0)
                    joined.append(undef + [right])
        elif j.get("right"):
            for right in j["items"]:
                m = False
                for left in items:
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    joined.append(left + [right])
                if not m:
                    undef = [None] * (len(items[0]) if items else 0)
                    joined.append(undef + [right])
        else:
            for left in items:
                m = False
                for right in j["items"]:
                    keep = True
                    if j.get("on"):
                        keep = j["on"](*left, right)
                    if not keep:
                        continue
                    m = True
                    joined.append(left + [right])
                if j.get("left") and (not m):
                    joined.append(left + [None])
        items = joined
    if opts.get("where"):
        items = [r for r in items if opts["where"](*r)]
    if opts.get("sortKey"):

        def _key(it):
            k = opts["sortKey"](*it)
            if isinstance(k, (list, tuple, dict)):
                return str(k)
            return k

        items.sort(key=_key)
    if "skip" in opts:
        n = opts["skip"]
        if n < 0:
            n = 0
        items = items[n:] if n < len(items) else []
    if "take" in opts:
        n = opts["take"]
        if n < 0:
            n = 0
        items = items[:n] if n < len(items) else items
    res = []
    for r in items:
        res.append(opts["select"](*r))
    return res


def test_Q5_finds_the_lexicographically_first_qualifying_title():
    assert result == [Auto1(typical_european_movie="A Film")]


company_type = [
    Auto2(ct_id=1, kind="production companies"),
    Auto2(ct_id=2, kind="other"),
]
info_type = [Auto3(it_id=10, info="languages")]
title = [
    Auto4(t_id=100, title="B Movie", production_year=2010),
    Auto4(t_id=200, title="A Film", production_year=2012),
    Auto4(t_id=300, title="Old Movie", production_year=2000),
]
movie_companies = [
    Auto5(movie_id=100, company_type_id=1, note="ACME (France) (theatrical)"),
    Auto5(movie_id=200, company_type_id=1, note="ACME (France) (theatrical)"),
    Auto5(movie_id=300, company_type_id=1, note="ACME (France) (theatrical)"),
]
movie_info = [
    Auto6(movie_id=100, info="German", info_type_id=10),
    Auto6(movie_id=200, info="Swedish", info_type_id=10),
    Auto6(movie_id=300, info="German", info_type_id=10),
]
candidate_titles = _query(
    company_type,
    [
        {
            "items": movie_companies,
            "on": lambda ct, mc: mc["company_type_id"] == ct["ct_id"],
        },
        {
            "items": movie_info,
            "on": lambda ct, mc, mi: mi["movie_id"] == mc["movie_id"],
        },
        {
            "items": info_type,
            "on": lambda ct, mc, mi, it: it["it_id"] == mi["info_type_id"],
        },
        {"items": title, "on": lambda ct, mc, mi, it, t: t["t_id"] == mc["movie_id"]},
    ],
    {
        "select": lambda ct, mc, mi, it, t: t["title"],
        "where": lambda ct, mc, mi, it, t: (
            (
                (ct["kind"] == "production companies" and "(theatrical)" in mc["note"])
                and "(France)" in mc["note"]
            )
            and t["production_year"] > 2005
        )
        and mi["info"]
        in [
            "Sweden",
            "Norway",
            "Germany",
            "Denmark",
            "Swedish",
            "Denish",
            "Norwegian",
            "German",
        ],
    },
)
result = [Auto1(typical_european_movie=_min(candidate_titles))]
print(json.dumps(result, default=lambda o: vars(o)))
test_Q5_finds_the_lexicographically_first_qualifying_title()